### PR TITLE
Add encryption option to backup script

### PR DESCRIPTION
**Pull request by fefi42**
I extended the backup script with an encryption option.
I also added two scripts one for decrypting a backup and one to generate an encryption key.

### CODE REVIEW

**Comment by fefi42** [comment]
The encryption is fernet from the `pyca/cryptography` package.

**Comment by SamLubbers** [comment]
@fefi42 great job. Like we discussed last meeting, could you replace the symmetric encryption with asymmetric encryption so that the encryption key cannot be found on the server?

In addition, could document the process for storing the encryption key? Ideally document this in the README. If that isn't safe it would have to be documented elsewhere.

**Comment by fefi42** [comment]
@SamLubbers I updated the code to use asymmetric encryption. The Backup now generates a zip file with the encrypted data. I extended the readme to explain how to store the encryption keys. Please let me know if anything is unclear.

**Comment by SamLubbers** [comment]
@fefi42 thank you for the changes. The instructions for operating the script and storing the keys are very clear 👌 

#### Simplify encryption
I think the encryption can be slightly simplified.

##### Current version
If I am correct, the current program does the following:

**Encryption:**
1. Encrypt the data with a symmetric key (`Fernet`)
2. Encrypt the symmetric key with the asymmetric public key
3. Store data (encrypted with symmetric key) and symmetric key (encrypted with the asymmetric public key) into a zip file

**Decryption:** 
1. Decrypt symmetric key using the asymmetric private key
2. Decrypt data using symmetric key.

##### Simplified version
I think we can have a simplified version without the symmetric encryption step:

**Encryption:** 
1. Encrypt the data with the asymmetric public key
2. Store encrypted data

**Decryption:** 
1. Decrypt data using the asymmetric private key

File: backup_script.py [hunk 1]
```diff
@@ -1,20 +1,108 @@
...
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import padding
+from rebel_management_utilities import get_all_members
+from cryptography.fernet import Fernet
+from zipfile import ZipFile
 
+def encrypt_data(data, asymmetric_public_key):
+    """ The data is encrypted using a symmetric key that will be encrypted by the assymetric public key.
+
+    :param key: The public key used to encrypt the symmetric key
```
Comment (SamLubbers) [inline backup_script.py]: documentation must match the argument name. If the use of the argument is obvious from its name it is not necessary to document it.

File: backup_script.py [hunk 2]
```diff
@@ -1,20 +1,108 @@
...
+
+
+
+
+def write_to_file_raw(data):
+    """ Write string data to file
+
+    :param data:
+    :return:
+    """
```
Comment (SamLubbers) [inline backup_script.py]: remove empty docs

File: backup_script.py [hunk 3]
```diff
@@ -1,20 +1,108 @@
...
 
-def make_backup(api_key):
-    data = {'members': get_all_members(api_key)}
     current_date_str = datetime.now().strftime("%d-%m-%Y_%H:%M:%S")
 
-    file_path = path.join('backups', f'backup_rebels_{current_date_str}.json')
+    backup_dir = 'backups'
```
Comment (SamLubbers) [inline backup_script.py]: `backup_dir = 'backups'` is declared twice. Once here and once in `write_to_file_encrypted`. You can avoid this duplication by moving the declaration outside of the functions.

File: decrypt_file.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from datetime import datetime
+import zipfile
+from cryptography.fernet import Fernet
+
+if __name__ == '__main__':
+    if len(sys.argv) != 3:
+        print("Please specify the path to the encryption key file and the path to the data zip file")
+        exit(1)
+
+    private_key = ""
```
Comment (SamLubbers) [inline decrypt_file.py]: This declaration is not necessary in Python
```suggestion
```

File: decrypt_file.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            password=None,
+            backend=default_backend()
+        )
+
+    # Open the zip archive
+    archive = zipfile.ZipFile(sys.argv[2], 'r')
+    try:
+        archive_content_name_list = archive.namelist()
+        symmetric_key_file = ''
+        data_file_encrypted = ''
```
Comment (SamLubbers) [inline decrypt_file.py]: declarations before if statements are not necessary in Python

**Comment by fefi42** [comment]
Hi @SamLubbers 

Encrypting the entire data with an asymmetric encryption is not possible because of the size of the data. To encrypt a file with the size of the backup we would need to generate a gigantically big key pair that would take hours to generate. Using the asymmetric key to encrypt the symmetric key is a common pattern to do this. 
Here is a stack exchange discussion that describes the benefits: https://security.stackexchange.com/a/53643

File: backup_script.py [hunk 6]
```diff
@@ -1,20 +1,108 @@
...
+        f.close()
+
+
+def write_to_file_encrypted(data):
+    """ Write data to file
+
+    :param data: Contains the symmetric key and the data
+    :type data: dict
+    :param is_encrypted: if true writes a binary file
+    :type bool
```
Comment (SamLubbers) [inline backup_script.py]: parameter does not exist in function declaration.

File: backup_script.py [hunk 7]
```diff
@@ -26,7 +114,34 @@ def load_api_key():
 
     return key
 
+def load_data(api_key):
+    data = {'members': get_all_members(api_key)}
+    return json.dumps(data)
```
Comment (SamLubbers) [inline backup_script.py]: Include these parameters for nicer file formatting. 
```suggestion
    return json.dumps(data, ensure_ascii=False, indent=4)
```

**Comment by SamLubbers** [comment]
@fefi42 thank you for the clarification. The pull request is ready to be merged once you resolve the last few changes I suggested.

**Pull request merged**